File: nutrition_ai/streaks.py
"""
Streak Engine — مبني على سجل أيام فعلية (ActiveDay) بدل عداد هش. توقيت بغداد دائمًا
(iraq_time.py)، ونشاط واحد أو أكثر بنفس اليوم = يوم واحد بالستريك (UniqueConstraint على
user_id+date يمنع المضاعفة تلقائيًا).
"""
from datetime import timedelta

import iraq_time
from models import db, ActiveDay, StreakMilestone
from nutrition_ai import xp_engine


DEFAULT_MILESTONES = [
    (1, 5, "أول يوم 🔥"),
    (3, 10, "3 أيام متتالية"),
    (7, 20, "أسبوع كامل 🔥"),
    (14, 35, "أسبوعين"),
    (30, 75, "شهر كامل 💪"),
    (60, 120, "شهرين"),
    (100, 200, "100 يوم 🏆"),
    (365, 500, "سنة كاملة 🏆"),
]


def seed_default_milestones():
    if StreakMilestone.query.first():
        return
    for days, xp_reward, label in DEFAULT_MILESTONES:
        db.session.add(StreakMilestone(days=days, xp_reward=xp_reward, label=label, active=True))
    db.session.commit()


def record_active_day(user) -> dict:
    """
    يسجّل اليوم الحالي (بغداد) كيوم نشط لو لسا ما انسجل، ويحدّث streak_days/longest_streak/
    streak_started_at تبع المستخدم. يرجّع Snapshot كامل (لأغراض التراجع عبر undo_active_day)
    + أي محطات (Milestones) جديدة تحققت بهذا التسجيل.
    """
    today = iraq_time.now_baghdad().date()
    snapshot = {
        "is_new_day": False, "active_day_id": None,
        "streak_before": user.streak_days, "longest_before": user.longest_streak,
        "streak_started_before": user.streak_started_at, "last_active_before": user.last_active_date,
    }

    existing = ActiveDay.query.filter_by(user_id=user.id, date=today).first()
    if existing:
        snapshot["active_day_id"] = existing.id
        return {**snapshot, "new_milestones": []}

    row = ActiveDay(user_id=user.id, date=today)
    db.session.add(row)
    db.session.flush()  # نحتاج row.id فورًا لأغراض التراجع

    yesterday = today - timedelta(days=1)
    had_yesterday = ActiveDay.query.filter_by(user_id=user.id, date=yesterday).first() is not None
    if had_yesterday:
        user.streak_days = snapshot["streak_before"] + 1
    else:
        user.streak_days = 1
        user.streak_started_at = today

    user.last_active_date = today
    user.longest_streak = max(snapshot["longest_before"], user.streak_days)

    new_milestones = _award_milestones(user, user.streak_days)

    return {
        **snapshot, "is_new_day": True, "active_day_id": row.id, "new_milestones": new_milestones,
    }


def _award_milestones(user, current_streak: int) -> list[dict]:
    milestones = StreakMilestone.query.filter(
        StreakMilestone.active.is_(True), StreakMilestone.days <= current_streak,
    ).all()
    newly_awarded = []
    for m in milestones:
        granted = xp_engine.award_xp(
            user, m.xp_reward, reason="streak_milestone", source=f"streak_milestone_{m.days}",
        )
        if granted:
            newly_awarded.append({"days": m.days, "label": m.label, "xp_reward": m.xp_reward})
    return newly_awarded


def undo_active_day(user, snapshot: dict) -> None:
    """يرجّع كل شي لقيمته قبل record_active_day — يُستخدم من direct_log.py عند تراجع/إعادة فتح."""
    if not snapshot.get("is_new_day"):
        return  # كان يوم نشط مسبقًا أصلاً — ما فيه شي نرجّعه
    if snapshot.get("active_day_id"):
        row = ActiveDay.query.get(snapshot["active_day_id"])
        if row:
            db.session.delete(row)
    user.streak_days = snapshot["streak_before"]
    user.longest_streak = snapshot["longest_before"]
    user.streak_started_at = snapshot["streak_started_before"]
    user.last_active_date = snapshot["last_active_before"]


def days_absent(user) -> int:
    """عدد الأيام منذ آخر نشاط. 0 لو نشط اليوم نفسه أو ما بدأ نشاط أبدًا."""
    if not user.last_active_date:
        return 0
    today = iraq_time.now_baghdad().date()
    return max(0, (today - user.last_active_date).days)


def recent_active_dates(user, days: int = 30) -> set:
    """آخر N يوم من ActiveDay — لعرض Calendar البروفايل."""
    today = iraq_time.now_baghdad().date()
    start = today - timedelta(days=days - 1)
    rows = ActiveDay.query.filter(
        ActiveDay.user_id == user.id, ActiveDay.date >= start, ActiveDay.date <= today,
    ).all()
    return {r.date for r in rows}

File: shared/nutrition-engine/streaks.ts
/**
 * منفذ حرفي من nutrition_ai/streaks.py — Streak Engine مبني على سجل أيام فعلية (ActiveDay)
 * بدل عداد هش. توقيت بغداد دائمًا، ونشاط واحد أو أكثر بنفس اليوم = يوم واحد بالستريك (قيد
 * فريد على user_id+date يمنع المضاعفة تلقائيًا على مستوى الـRepository).
 */
import type { Repository, UserRecord } from "./db/repository.js";
import * as xpEngine from "./xpEngine.js";
import { todayBaghdadIso, addDaysIso, diffDaysIso } from "./iraqTime.js";

export interface MilestoneAwarded {
  days: number;
  label: string;
  xp_reward: number;
}

export interface StreakSnapshot {
  is_new_day: boolean;
  active_day_id: string | null;
  streak_before: number;
  longest_before: number;
  streak_started_before: string | null;
  last_active_before: string | null;
  new_milestones: MilestoneAwarded[];
}

/**
 * يسجّل اليوم الحالي (بغداد) كيوم نشط لو لسا ما انسجل، ويحدّث streak_days/longest_streak/
 * streak_started_at تبع المستخدم. يرجّع Snapshot كامل (لأغراض التراجع عبر undoActiveDay) + أي
 * محطات (Milestones) جديدة تحققت بهذا التسجيل.
 */
export async function recordActiveDay(repo: Repository, user: UserRecord, now: Date = new Date()): Promise<StreakSnapshot> {
  const today = todayBaghdadIso(now);
  const snapshot: Omit<StreakSnapshot, "new_milestones"> = {
    is_new_day: false, active_day_id: null,
    streak_before: user.streak_days, longest_before: user.longest_streak,
    streak_started_before: user.streak_started_at, last_active_before: user.last_active_date,
  };

  const existing = await repo.findActiveDay(user.id, today);
  if (existing) {
    return { ...snapshot, active_day_id: existing.id, new_milestones: [] };
  }

  const row = await repo.insertActiveDay(user.id, today);

  const yesterday = addDaysIso(today, -1);
  const hadYesterday = (await repo.findActiveDay(user.id, yesterday)) !== null;
  if (hadYesterday) {
    user.streak_days = snapshot.streak_before + 1;
  } else {
    user.streak_days = 1;
    user.streak_started_at = today;
  }

  user.last_active_date = today;
  user.longest_streak = Math.max(snapshot.longest_before, user.streak_days);

  const newMilestones = await awardMilestones(repo, user, user.streak_days);

  return { ...snapshot, is_new_day: true, active_day_id: row.id, new_milestones: newMilestones };
}

async function awardMilestones(repo: Repository, user: UserRecord, currentStreak: number): Promise<MilestoneAwarded[]> {
  const milestones = await repo.listActiveStreakMilestonesUpTo(currentStreak);
  const newlyAwarded: MilestoneAwarded[] = [];
  for (const m of milestones) {
    const granted = await xpEngine.awardXp(repo, user, m.xp_reward, "streak_milestone", `streak_milestone_${m.days}`);
    if (granted) {
      newlyAwarded.push({ days: m.days, label: m.label, xp_reward: m.xp_reward });
    }
  }
  return newlyAwarded;
}

/** يرجّع كل شي لقيمته قبل recordActiveDay — يُستخدم من directLog.ts عند تراجع/إعادة فتح. */
export async function undoActiveDay(repo: Repository, user: UserRecord, snapshot: StreakSnapshot): Promise<void> {
  if (!snapshot.is_new_day) return; // كان يوم نشط مسبقًا أصلاً — ما فيه شي نرجّعه
  if (snapshot.active_day_id) {
    const row = await repo.findActiveDayById(snapshot.active_day_id);
    if (row) await repo.deleteActiveDay(row.id);
  }
  user.streak_days = snapshot.streak_before;
  user.longest_streak = snapshot.longest_before;
  user.streak_started_at = snapshot.streak_started_before;
  user.last_active_date = snapshot.last_active_before;
}

/** عدد الأيام منذ آخر نشاط. 0 لو نشط اليوم نفسه أو ما بدأ نشاط أبدًا. */
export function daysAbsent(user: UserRecord, now: Date = new Date()): number {
  if (!user.last_active_date) return 0;
  const today = todayBaghdadIso(now);
  return Math.max(0, diffDaysIso(today, user.last_active_date));
}

/** آخر N يوم من ActiveDay — لعرض Calendar البروفايل. */
export async function recentActiveDates(repo: Repository, user: UserRecord, days = 30, now: Date = new Date()): Promise<Set<string>> {
  const today = todayBaghdadIso(now);
  const start = addDaysIso(today, -(days - 1));
  const dates = await repo.findActiveDaysInRange(user.id, start, today);
  return new Set(dates);
}
